rules/terraform_unused_declarations.go

```diff
@@ -138,17 +138,38 @@ func (r *TerraformUnusedDeclarationsRule) declarations(runner *terraform.Runner)
 				LabelNames: []string{"type", "name"},
 				Body:       &hclext.BodySchema{},
 			},
+			{
+				Type:       "check",
+				LabelNames: []string{"name"},
+				Body: &hclext.BodySchema{
+					Blocks: []hclext.BlockSchema{
+						{
+							Type:       "data",
+							LabelNames: []string{"type", "name"},
+							Body:       &hclext.BodySchema{},
+						},
+					},
+				},
+			},
 		},
 	}, &tflint.GetModuleContentOption{ExpandMode: tflint.ExpandModeNone})
 	if err != nil {
 		return decl, err
 	}
 
 	for _, block := range body.Blocks {
-		if block.Type == "variable" {
+		switch block.Type {
+		case "variable":
 			decl.Variables[block.Labels[0]] = block
-		} else {
+		case "data":
 			decl.DataResources[fmt.Sprintf("data.%s.%s", block.Labels[0], block.Labels[1])] = block
+		case "check":
+			for _, data := range block.Body.Blocks {
+				// Scoped data source addresses are unique in the module
+				decl.DataResources[fmt.Sprintf("data.%s.%s", data.Labels[0], data.Labels[1])] = data
```

Comment:
Note: This logic is broken if scoped data source addresses are unique in `check` blocks.

As far as I've looked, this logic is valid because in v1.6.0 it's unique in the module.

```hcl
data "http" "terraform_io" {
  url = "https://www.terraform.io"
}

check "health_check" {
  data "http" "terraform_io" {
    url = "https://www.terraform.io"
  }

  assert {
    condition = data.http.terraform_io.status_code == 200
    error_message = "${data.http.terraform_io.url} returned an unhealthy status code"
  }
}
```

```
@wata727 ➜ /tmp $ terraform validate
╷
│ Error: Duplicate data "http" configuration
│
│   on main.tf line 6, in check "health_check":
│    6:   data "http" "terraform_io" {
│
│ A http data resource named "terraform_io" was already declared at main.tf:1,1-27. Resource names must be unique per type in each module, including within check
│ blocks.
```